src/edge_utils.c

```diff
@@ -2344,7 +2358,7 @@ int run_edge_loop(n2n_edge_t * eee, int *keep_running) {
     update_supernode_reg(eee, nowTime);
 
     numPurged =  purge_expired_registrations(&eee->known_peers, &last_purge_known, PURGE_REGISTRATION_FREQUENCY);
-    numPurged += purge_expired_registrations(&eee->pending_peers, &last_purge_pending, PURGE_REGISTRATION_FREQUENCY);
+    numPurged += purge_expired_registrations(&eee->pending_peers, &last_purge_pending, eee->conf.register_interval);
```

Comment:
I do not fully understand the reason for this change. Could you please provide a short explanation?

Reply:
I ran into a situation:  After run some time, `edge A` and `edge B` always communicate by supernode forward. But if I kill and restart any edge, `edge A` and `edge B` can punching success.

I find that `MSG_TYPE_PACKET` and `MSG_TYPE_REGISTER` will call `check_peer_registration_needed`, if punching not success, peer not exists in `known_peers`, `register_with_new_peer` will be called. Inside `register_with_new_peer`,  if peer already in `pending_peers`, nothing will do except update `last_seen`.
 
What if B is in A's `pending_peers`, same time A is in B's  `pending_peers`, but punching temporary failed? 

In summary, current code, punching will execute only once when add peer to `pending_peers`. If some reason this time punching failed, `last_seen` will be update every `MSG_TYPE_PACKET` and `MSG_TYPE_REGISTER` packet. So purge_expired_registrations cannot clear that peer from `pending_peers` because it use `last_seen`. So no punching retry will happen.


So I made a temporary imperfections fix: only update `last_seen` at `MSG_TYPE_PACKET`, at my code, this type packet is dropped by `-P` if not punching success.  So if punching not success for some time, peer will be delete from `pending_peers`. Then next time when received a packet from a peer forward by supernode, `check_peer_registration_needed` will tigger a new punching process while add peer to `pending_peers`.

`is_packet_msg` is used to pass info (packet type is `MSG_TYPE_PACKET`) to functions which be called.

This is not a perfect fix for that issue, maybe we need find an other way.

Reply:
This situation is found before I introduce the `-P` function.  And it can be confirm at dev@ntop/n2n at this time.

Reply:
I have made a new way to let punching retry:

When begin punching (add peer to pending_peers), record time into `last_try_punching_time`.

At `run_edge_loop`, use `purge_expired_pending_peers` instead of `purge_expired_registrations` to purge pending peers by last_try_punching_time.

All changes about `is_packet_msg` and `last_seen` is reverted.

==== working flow description ====

When punching started, record the time. There has 10s to wait punching success.
If success, peer will be delete from `pending_peers`, insert into `known_peers`. According to current upstream codes.
If not success, peer will be delete from `pending_peers` by `purge_expired_pending_peers` because of timeout(10s later).
Once peer be deleted from `pending_peers`, it will be treat as new peer, and a new punching process will be trigger when MSG_TYPE_PACKET or MSG_TYPE_REGISTER is received from that peer.

Reply:
@Logan007 What is your opinion? Are you happy with this?

Reply:
Basically, I think you have provided n2n with an excellent idea @joshuafc and I absolutely support its implementation.

I still am struggling with some details which might be caused by a lack of knowledge about networks on my side, however, the following questions arise – maybe, answers are easy:

- Is it really likely that punching fails, i.e. that the fields `last_seen` and `last_try_punching_time` fall apart? Do we really need that other field?

- **If** this _only-ARP-through-the-supernode_ approach keeps the network working, I see no issue in not implementing it. Does the `known_peers` entry (at the edge) or the `peer_info` entry (at the supernode, eventually asked for socket by MAC through QUERY_PEER_INFO mechanism) time-wise survive the ARP entry? What happens if connection shall be built up while those n2n-entries have gotten purged and ARP entry is still in table?

- What happens if  `-S` and `-P` are used parallel? Both options control the behavior of PACKET (and maybe REGISTER and REGISTER_ACK) sending and forwarding. Shouldn't we help the user to avoid this situation (maybe check if both options are present)? A future extension (in a separate pull request, not now) could merge  these two options to one `-S_` with, for example, `-S0` indicating the (default) regular behavior, `-S1` the current `-S`, `-S2` the current `-P`.

- Does the supernode really requires the `-P` option? It affects all communities then. Maybe you had different use-cases in mind, but wouldn't it be sufficient to leave it to the edges only to use `-P` or not `-P`? If they do not send out anything, the supernode does not need to reject anything.

Reply:
- `last_seen` will be update if received a supernode forward `PACKET` or `REGISTER`.

This implementation currently instability. As I Mentioned, codes at `ntop:dev` branch:

> After run some time, edge A and edge B always communicate by supernode forward(It cost me a lot of traffic bills). But if I kill and restart both edge, 
> edge A and edge B can punching success. 

Although I try add a re-punching by `last_try_punching_time`, this will still happen. With `-P`, edge A and edge B link broken forever.

Maybe we need to solve the re-punching problem first, It's a little hard for me to solve the punching issue. After that, I will make a new edge-only implementation.